ssl/quic/quic_lcidm.c

```diff
@@ -0,0 +1,519 @@
+                             void *opaque,
+                             OSSL_QUIC_FRAME_NEW_CONN_ID *ncid_frame)
+{
+    ncid_frame->seq_num         = 0;
+    ncid_frame->retire_prior_to = 0;
+
+    return lcidm_generate(lcidm, opaque, LCID_TYPE_NCID,
+                          &ncid_frame->conn_id,
+                          &ncid_frame->seq_num);
+}
+
+int ossl_quic_lcidm_retire_odcid(QUIC_LCIDM *lcidm, void *opaque)
+{
+    QUIC_LCIDM_CONN *conn;
+
+    if ((conn = lcidm_upsert_conn(lcidm, opaque)) == NULL)
+        return 0;
+
+    if (conn->odcid_lcid == NULL)
+        return 0;
+
+    lcidm_delete_conn_lcid(lcidm, conn->odcid_lcid);
+    conn->odcid_lcid = NULL;
+    return 1;
+}
+
+struct retire_args {
+    QUIC_LCID           *earliest_seq_num_lcid;
+    uint64_t            earliest_seq_num, retire_prior_to;
+};
+
+static void retire_for_conn(QUIC_LCID *lcid, void *arg)
+{
+    struct retire_args *args = arg;
+
+    /* ODCID LCID cannot be retired via this API */
+    if (lcid->type == LCID_TYPE_ODCID
+        || lcid->seq_num >= args->retire_prior_to)
+        return;
+
+    if (lcid->seq_num < args->earliest_seq_num) {
+        args->earliest_seq_num = lcid->seq_num;
+        args->earliest_seq_num_lcid = lcid;
+    }
+}
+
+int ossl_quic_lcidm_retire(QUIC_LCIDM *lcidm,
+                           void *opaque,
+                           uint64_t retire_prior_to,
+                           const QUIC_CONN_ID *containing_pkt_dcid,
+                           QUIC_CONN_ID *retired_lcid,
+                           uint64_t *retired_seq_num,
+                           int *did_retire)
+{
+    QUIC_LCIDM_CONN key, *conn;
+    struct retire_args args = {0};
+
+    key.opaque = opaque;
+
+    if (did_retire == NULL)
+        return 0;
+
+    *did_retire = 0;
+    if ((conn = lh_QUIC_LCIDM_CONN_retrieve(lcidm->conns, &key)) == NULL)
+        return 1;
+
+    args.retire_prior_to    = retire_prior_to;
+    args.earliest_seq_num   = UINT64_MAX;
+    lh_QUIC_LCID_doall_arg(conn->lcids, retire_for_conn, &args);
+    if (args.earliest_seq_num_lcid == NULL)
+        return 1;
+
+    if (containing_pkt_dcid != NULL
+        && ossl_quic_conn_id_eq(&args.earliest_seq_num_lcid->cid,
+                                containing_pkt_dcid))
+        return 0;
+
+    *did_retire = 1;
+    if (retired_lcid != NULL)
+        *retired_lcid = args.earliest_seq_num_lcid->cid;
+    if (retired_seq_num != NULL)
+        *retired_seq_num = args.earliest_seq_num_lcid->seq_num;
+
+    lcidm_delete_conn_lcid(lcidm, args.earliest_seq_num_lcid);
+    return 1;
+}
+
+int ossl_quic_lcidm_cull(QUIC_LCIDM *lcidm, void *opaque)
+{
+    QUIC_LCIDM_CONN key, *conn;
+
+    key.opaque = opaque;
+
+    if ((conn = lh_QUIC_LCIDM_CONN_retrieve(lcidm->conns, &key)) == NULL)
+        return 0;
+
+    lcidm_delete_conn(lcidm, conn);
+    return 1;
+}
+
+int ossl_quic_lcidm_lookup(QUIC_LCIDM *lcidm,
+                           const QUIC_CONN_ID *lcid,
```

Comment:
I find this use of the name `lcid` quite confusing (from the name I expect the type to be `QUIC_LCID`). Here and in other functions